plutus-emulator/src/Wallet/Emulator/MultiAgent.hs

```diff
@@ -0,0 +1,220 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE Rank2Types            #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TupleSections         #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# OPTIONS_GHC -Wno-overlapping-patterns #-}
+module Wallet.Emulator.MultiAgent where
+
+import           Control.Lens
+import           Control.Monad
+import           Control.Monad.Freer
+import           Control.Monad.Freer.Error
+import           Control.Monad.Freer.Extras
+import           Control.Monad.Freer.State
+import           Data.Aeson                 (FromJSON, ToJSON)
+import           Data.Map                   (Map)
+import qualified Data.Map                   as Map
+import qualified Data.Text                  as T
+import           Data.Text.Prettyprint.Doc  hiding (annotate)
+import           GHC.Generics               (Generic)
+import           Ledger                     hiding (value)
+import qualified Ledger.Index               as Index
+import qualified Wallet.API                 as WAPI
+import qualified Wallet.Emulator.Chain      as Chain
+import qualified Wallet.Emulator.NodeClient as NC
+import qualified Wallet.Emulator.Wallet     as Wallet
+
+-- | Assertions which will be checked during execution of the emulator.
+data Assertion
+  = IsValidated Tx -- ^ Assert that the given transaction is validated.
+  | OwnFundsEqual Wallet.Wallet Value -- ^ Assert that the funds belonging to a wallet's public-key address are equal to a value.
+
+-- | An error emitted when an 'Assertion' fails.
+newtype AssertionError = GenericAssertion T.Text
+    deriving (Show, Eq)
+makeClassyPrisms ''AssertionError
+
+-- | This lets people use 'T.Text' as their error type.
+instance AsAssertionError T.Text where
```

Comment:
Yeah, we should be able to get rid of a lot of this in the future since we can hopefully just have multiple error effects instead of the prism approach.